Documentation/api-guidelines/nullability.md

```diff
@@ -0,0 +1,77 @@
+
+- **Mistakes.** Given the sheer number of APIs to be reviewed and annotated, we are likely to make some errors in annotations, and we'd like to be able to fix them.
+- **Breadth.** We are unlikely to have the time to annotate all of the enormous number of APIs in .NET for an initial release.
+- **Feedback.** We may need to revisit some "gray area" decisions as to whether a parameter or return type should be nullable or non-nullable.
+
+As such, while changing the annotations on APIs may impact warnings in the codebase of a developer that has opted-in to nullability warnings, for at least the foreseeable future we may still do so.  We will be very thoughtful about when and how we do so.
+
+## Annotation Guidance
+
+Nullability annotations are considered to represent intent: they represent the nullability contract for the member.  Any deviation from that intent on the part of an implementation should be considered an implementation bug, and the compiler will help to minimize the chances of such bugs via its flow analysis and nullability warnings. At the same time, it�s important to recognize that the validation performed by the compiler isn�t perfect; it can have both false positive warnings (suggesting that something may be null even when it isn�t) and false negatives (not warning when something that may be null is dereferenced).  The compiler cannot guarantee that an API declared as returning a non-nullable reference never returns null, just as it can�t validate that an implementation declared as accepting nulls always behaves correctly when given them.  When deciding how to annotate APIs, it�s important then to consider the desired contract rather than the current implementation; in other words, prefer to first annotate the API surface area the way that�s desired, and only then work to address any warnings in the codebase, rather than driving the API surface area annotations based on where those warnings lead.
+
+The majority of reference type usage in our APIs is fairly clear as to whether it should be nullable or not. For parameters, these general guidelines cover the majority of cases:
+
+- If the method checks for null and throws an ArgumentException if null is passed in for that parameter (whether explicitly in that same method or implicitly as part of some method it calls), such that there�s no way null could be passed in and have the method return successfully, the parameter should be declared as non-nullable.
+- If the method fails to check for null but instead will always end up dereferencing the null and throwing a NullReferenceException, the parameter should be declared as non-nullable.
+- If the parameter is explicitly documented to accept null, it should be declared nullable.
+- If the parameter explicitly checks for null and does something other than through, it should be declared nullable.  This may include normalizing the input, e.g. treating null as string.Empty.
```

Comment:
Through->throw

Reply:
```suggestion
- If the parameter explicitly checks for null and does something other than throw, it should be declared nullable.  This may include normalizing the input, e.g. treating null as string.Empty.
```